crates/rust-analyzer/src/config/tree.rs

```diff
@@ -0,0 +1,685 @@
+            self.computed_local_config(self.xdg_config_file_id).0
+        }
+    }
+
+    /// Applies a bunch of [`ConfigChanges`]. The FileIds referred to in `ConfigChanges` do not
+    /// need to exist.
+    pub fn apply_changes(&mut self, changes: ConfigChanges, vfs: &mut Vfs) -> Vec<ConfigTreeError> {
+        if let Some(new_project_root) = &changes.set_project_root {
+            self.project_root = new_project_root.clone();
+        }
+        let source_root_change = changes.set_source_roots.as_ref().or_else(|| {
+            if changes.set_project_root.is_some() {
+                Some(&self.source_roots)
+            } else {
+                None
+            }
+        });
+        let parent_changes = if let Some(source_roots) = source_root_change {
+            let parent_changes = source_roots
+                .iter()
+                .flat_map(|path: &AbsPathBuf| {
+                    path.ancestors()
+                        // Note that Path::new("/root2/abc").starts_with(Path::new("/root")) is false
+                        .take_while(|x| x.starts_with(&self.project_root))
+                        .map(|dir| dir.join("rust-analyzer.toml"))
+                        .map(|path| vfs.alloc_file_id(path.into()))
+                        .collect_vec()
+                        // immediately get tuple_windows before returning from flat_map
+                        .into_iter()
+                        .tuple_windows()
+                        .map(|(a, b)| (a, ConfigParent::Parent(b)))
+                })
+                .collect::<FxHashMap<_, _>>();
+
+            // Remove source roots (& their parent config files) that are no longer part of the project root
+            self.known_file_ids
+                .iter()
+                .cloned()
+                .filter(|&x| x != self.xdg_config_file_id && !parent_changes.contains_key(&x))
+                .collect_vec()
+                .into_iter()
+                .for_each(|deleted| {
+                    self.known_file_ids.remove(&deleted);
+                    self.reset_node(deleted);
+                });
+
+            parent_changes
+        } else {
+            Default::default()
+        };
+
+        if tracing::enabled!(tracing::Level::TRACE) {
+            for (&a, parent) in &parent_changes {
+                tracing::trace!(
+                    "{a:?} ({:?}): parent = {parent:?} ({:?})",
+                    vfs.file_path(a),
+                    match parent {
+                        ConfigParent::Parent(p) => vfs.file_path(*p).to_string(),
+                        ConfigParent::UserDefault => "xdg".to_string(),
+                    }
+                );
+            }
+        }
+
+        let inner = ConfigChangesInner {
+            ra_toml_changes: changes.ra_toml_changes,
+            client_change: changes.client_change,
+            parent_changes,
+        };
+        self.apply_changes_inner(inner, vfs)
+    }
+
+    fn apply_changes_inner(
```

Comment:
Why don't you have a single `apply_changes` instead of `apply_changes` + `apply_changes_inner`?

Reply:
I wrote apply_changes one on top of apply_changes_inner. This is just how the code evolved. The _inner one is completely free of any path information, works entirely on FileIds, which I think is nice, given the stress placed on being able to do that in the dev docs. I figured at some point we would make this code work for virtual paths as well as AbsPathBufs, and on that day, apply_changes_inner will remain unchanged.

Reply:
Notably, `apply_changes_inner` does not require `&mut Vfs`. So it helps isolate that unsavoury part as well, and you are welcome to split it off if you want to put the Vfs mutation closer to the other Vfs mutations that we already do.